File: app_pojav_zh/src/main/java/net/kdt/pojavlaunch/fragments/ControlButtonFragment.java
package net.kdt.pojavlaunch.fragments;

import static net.kdt.pojavlaunch.CustomControlsActivity.BUNDLE_CONTROL_PATH;
import static net.kdt.pojavlaunch.PojavZHTools.copyFileInBackground;
import static net.kdt.pojavlaunch.Tools.runOnUiThread;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.ipaulpro.afilechooser.FileIcon;
import com.kdt.pickafile.FileListView;
import com.kdt.pickafile.FileSelectedListener;

import net.kdt.pojavlaunch.CustomControlsActivity;
import net.kdt.pojavlaunch.PojavApplication;
import net.kdt.pojavlaunch.PojavZHTools;
import net.kdt.pojavlaunch.R;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.contracts.OpenDocumentWithExtension;
import net.kdt.pojavlaunch.dialog.CopyDialog;
import net.kdt.pojavlaunch.dialog.DeleteDialog;
import net.kdt.pojavlaunch.dialog.EditTextDialog;
import net.kdt.pojavlaunch.dialog.FilesDialog;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;

public class ControlButtonFragment extends Fragment {
    public static final String TAG = "ControlButtonFragment";
    private ActivityResultLauncher<Object> openDocumentLauncher;
    private Button mReturnButton, mAddControlButton, mImportControlButton, mRefreshButton;
    private ImageButton mHelpButton;
    private FileListView mFileListView;
    private TextView mFilePathView;

    public ControlButtonFragment() {
        super(R.layout.fragment_files);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        openDocumentLauncher = registerForActivityResult(
                new OpenDocumentWithExtension("json"),
                result -> {
                    if (result != null) {
                        Toast.makeText(requireContext(), getString(R.string.tasks_ongoing), Toast.LENGTH_SHORT).show();

                        PojavApplication.sExecutorService.execute(() -> {
                            copyFileInBackground(requireContext(), result, mFileListView.getFullPath().getAbsolutePath());

                            runOnUiThread(() -> {
                                Toast.makeText(requireContext(), getString(R.string.zh_file_added), Toast.LENGTH_SHORT).show();
                                mFileListView.refreshPath();
                            });
                        });
                    }
                }
        );
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        bindViews(view);

        mFileListView.setShowFiles(true);
        mFileListView.setShowFolders(true);
        mFileListView.lockPathAt(controlPath());
        mFileListView.setDialogTitleListener((title) -> mFilePathView.setText(removeLockPath(title)));
        mFileListView.refreshPath();

        mFileListView.setFileSelectedListener(new FileSelectedListener() {
            @Override
            public void onFileSelected(File file, String path) {
                FilesDialog.FilesButton filesButton = new FilesDialog.FilesButton();
                filesButton.setButtonVisibility(true, true, true, true);

                filesButton.messageText = getString(R.string.zh_file_message) + "\n" + getString(R.string.zh_file_message_copy);
                filesButton.moreButtonText = getString(R.string.global_load);

                FilesDialog filesDialog = new FilesDialog(requireContext(), filesButton, () -> runOnUiThread(() -> mFileListView.refreshPath()), file);
                filesDialog.setMoreButtonClick(v -> {
                    Intent intent = new Intent(requireContext(), CustomControlsActivity.class);

                    Bundle bundle = new Bundle();
                    bundle.putString(BUNDLE_CONTROL_PATH, file.getAbsolutePath());
                    intent.putExtras(bundle);

                    startActivity(intent);
                    filesDialog.dismiss();
                }); //加载
                filesDialog.show();
            }

            @Override
            public void onItemLongClick(File file, String path) {
                if (file.isDirectory()) {
                    DeleteDialog deleteDialog = new DeleteDialog(requireContext(), () -> runOnUiThread(() -> mFileListView.refreshPath()), file);
                    deleteDialog.show();
                } else {
                    CopyDialog dialog = new CopyDialog(requireContext(), mFileListView, file);
                    dialog.show();
                }
            }
        });

        mReturnButton.setOnClickListener(v -> PojavZHTools.onBackPressed(requireActivity()));
        mImportControlButton.setOnClickListener(v -> {
            String suffix = ".json";
            Toast.makeText(requireActivity(), String.format(getString(R.string.zh_file_add_file_tip), suffix), Toast.LENGTH_SHORT).show();
            openDocumentLauncher.launch(suffix);
        }); //限制.json文件
        mAddControlButton.setOnClickListener(v -> {
            EditTextDialog editTextDialog = new EditTextDialog(requireContext(), getString(R.string.zh_controls_create_new_title), null, null, null);
            editTextDialog.setConfirm(view1 -> {
                File file = new File(mFileListView.getFullPath().getAbsolutePath(), editTextDialog.getEditBox().getText().toString().replace("/", "") + ".json");
                if (!file.exists()) {
                    boolean success;
                    try {
                        success = file.createNewFile();
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                    if (success) {
                        try (BufferedWriter optionFileWriter = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8))) {
                            optionFileWriter.write("{\"version\":6}");
                        } catch (IOException e) {
                            Tools.showError(requireContext(), e);
                        }
                    }
                    mFileListView.refreshPath();
                } else {
                    Toast.makeText(requireContext(), getString(R.string.zh_file_create_file_invalid), Toast.LENGTH_SHORT).show();
                }

                editTextDialog.dismiss();
            });
            editTextDialog.show();
        });
        mRefreshButton.setOnClickListener(v -> mFileListView.refreshPath());
        mHelpButton.setOnClickListener(v -> {
            AlertDialog.Builder builder = new AlertDialog.Builder(requireActivity());

            builder.setTitle(getString(R.string.zh_help_control_button_title));
            builder.setMessage(getString(R.string.zh_help_control_button_message));
            builder.setPositiveButton(getString(R.string.zh_help_ok), null);

            builder.show();
        });
    }

    private File controlPath() {
        File ctrlPath = new File(Tools.CTRLMAP_PATH);
        if (!ctrlPath.exists()) ctrlPath.mkdirs();
        return ctrlPath;
    }

    private String removeLockPath(String path) {
        return path.replace(Tools.CTRLMAP_PATH, ".");
    }

    private void bindViews(@NonNull View view) {
        mReturnButton = view.findViewById(R.id.zh_files_return_button);
        mImportControlButton = view.findViewById(R.id.zh_files_add_file_button);
        mAddControlButton = view.findViewById(R.id.zh_files_create_folder_button);
        mRefreshButton = view.findViewById(R.id.zh_files_refresh_button);
        mHelpButton = view.findViewById(R.id.zh_files_help_button);
        mFileListView = view.findViewById(R.id.zh_files);
        mFilePathView = view.findViewById(R.id.zh_files_current_path);

        mImportControlButton.setText(getString(R.string.zh_controls_import_control));
        mAddControlButton.setText(getString(R.string.zh_controls_create_new));

        view.findViewById(R.id.zh_files_icon).setVisibility(View.GONE);
        mFileListView.setFileIcon(FileIcon.CONTROL);
    }
}

File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/fragment/ControlButtonFragment.kt
package com.movtery.pojavzh.ui.fragment

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import com.getkeepsafe.taptargetview.TapTargetSequence
import com.movtery.anim.AnimPlayer
import com.movtery.anim.animations.Animations
import com.movtery.pojavzh.setting.Settings
import com.movtery.pojavzh.ui.dialog.EditControlInfoDialog
import com.movtery.pojavzh.ui.dialog.FilesDialog
import com.movtery.pojavzh.ui.dialog.FilesDialog.FilesButton
import com.movtery.pojavzh.ui.dialog.TipDialog
import com.movtery.pojavzh.ui.subassembly.customcontrols.ControlInfoData
import com.movtery.pojavzh.ui.subassembly.customcontrols.ControlSelectedListener
import com.movtery.pojavzh.ui.subassembly.customcontrols.ControlsListViewCreator
import com.movtery.pojavzh.ui.subassembly.customcontrols.EditControlData.Companion.createNewControlFile
import com.movtery.pojavzh.ui.subassembly.view.SearchViewWrapper
import com.movtery.pojavzh.utils.NewbieGuideUtils
import com.movtery.pojavzh.utils.PathAndUrlManager
import com.movtery.pojavzh.utils.ZHTools
import com.movtery.pojavzh.utils.anim.AnimUtils.Companion.setVisibilityAnim
import com.movtery.pojavzh.utils.file.FileTools.Companion.copyFileInBackground
import com.movtery.pojavzh.utils.file.PasteFile
import net.kdt.pojavlaunch.CustomControlsActivity
import net.kdt.pojavlaunch.PojavApplication
import net.kdt.pojavlaunch.R
import net.kdt.pojavlaunch.Tools
import net.kdt.pojavlaunch.contracts.OpenDocumentWithExtension
import net.kdt.pojavlaunch.extra.ExtraConstants
import net.kdt.pojavlaunch.extra.ExtraCore
import java.io.File

class ControlButtonFragment : FragmentWithAnim(R.layout.fragment_control_manager) {
    companion object {
        const val TAG: String = "ControlButtonFragment"
        const val BUNDLE_SELECT_CONTROL: String = "bundle_select_control"
    }

    private var openDocumentLauncher: ActivityResultLauncher<Any>? = null
    private var mControlLayout: View? = null
    private var mOperateLayout: View? = null
    private var mOperateView: View? = null
    private var mReturnButton: ImageButton? = null
    private var mAddControlButton: ImageButton? = null
    private var mImportControlButton: ImageButton? = null
    private var mPasteButton: ImageButton? = null
    private var mSearchSummonButton: ImageButton? = null
    private var mRefreshButton: ImageButton? = null
    private var mNothingTip: TextView? = null
    private var mSearchViewWrapper: SearchViewWrapper? = null
    private var controlsListViewCreator: ControlsListViewCreator? = null
    private var mSelectControl = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        openDocumentLauncher = registerForActivityResult(OpenDocumentWithExtension("json")) { result: Uri? ->
            result?.let {
                Toast.makeText(requireContext(), getString(R.string.tasks_ongoing), Toast.LENGTH_SHORT).show()

                PojavApplication.sExecutorService.execute {
                    copyFileInBackground(requireContext(), result, File(PathAndUrlManager.DIR_CTRLMAP_PATH!!).absolutePath)
                    Tools.runOnUiThread {
                        Toast.makeText(requireContext(), getString(R.string.zh_file_added), Toast.LENGTH_SHORT).show()
                        controlsListViewCreator?.refresh()
                    }
                }
            }
        }
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        bindViews(view)
        parseBundle()

        controlsListViewCreator?.setSelectedListener(object : ControlSelectedListener() {
            override fun onItemSelected(file: File) {
                if (mSelectControl) {
                    ExtraCore.setValue(ExtraConstants.FILE_SELECTOR, removeLockPath(file.absolutePath))
                    Tools.removeCurrentFragment(requireActivity())
                } else {
                    if (file.isFile) showDialog(file)
                }
            }

            override fun onItemLongClick(file: File) {
                TipDialog.Builder(requireContext())
                    .setTitle(R.string.default_control)
                    .setMessage(R.string.zh_controls_set_default_message)
                    .setConfirmClickListener {
                        val absolutePath = file.absolutePath
                        Settings.Manager.put("defaultCtrl", absolutePath).save()
                    }.buildDialog()
            }
        })

        controlsListViewCreator?.setRefreshListener {
            val itemCount = controlsListViewCreator!!.itemCount
            val show = itemCount == 0
            setVisibilityAnim(mNothingTip!!, show)
        }

        mReturnButton?.setOnClickListener { ZHTools.onBackPressed(requireActivity()) }
        mPasteButton?.setOnClickListener {
            PasteFile.getInstance().pasteFiles(requireActivity(), File(PathAndUrlManager.DIR_CTRLMAP_PATH!!), null) {
                Tools.runOnUiThread {
                    mPasteButton?.visibility = View.GONE
                    controlsListViewCreator?.refresh()
                }
            }
        }
        mImportControlButton?.setOnClickListener {
            val suffix = ".json"
            Toast.makeText(requireActivity(), String.format(getString(R.string.zh_file_add_file_tip), suffix), Toast.LENGTH_SHORT).show()
            openDocumentLauncher?.launch(suffix)
        } //限制.json文件
        mAddControlButton?.setOnClickListener {
            val editControlInfoDialog = EditControlInfoDialog(requireContext(), true, null, ControlInfoData())
            editControlInfoDialog.setTitle(getString(R.string.zh_controls_create_new))
            editControlInfoDialog.setOnConfirmClickListener { fileName: String, controlInfoData: ControlInfoData? ->
                val file = File(File(PathAndUrlManager.DIR_CTRLMAP_PATH!!).absolutePath, "$fileName.json")
                if (file.exists()) { //检查文件是否已经存在
                    editControlInfoDialog.fileNameEditBox.error =
                        getString(R.string.zh_file_rename_exitis)
                    return@setOnConfirmClickListener
                }

                //创建布局文件
                createNewControlFile(requireContext(), file, controlInfoData)

                controlsListViewCreator?.refresh()
                editControlInfoDialog.dismiss()
            }
            editControlInfoDialog.show()
        }
        mSearchSummonButton?.setOnClickListener { mSearchViewWrapper?.setVisibility() }
        mRefreshButton?.setOnClickListener { controlsListViewCreator?.refresh() }

        controlsListViewCreator?.listAtPath()

        startNewbieGuide()
    }

    private fun removeLockPath(path: String?): String {
        return path!!.replace(PathAndUrlManager.DIR_CTRLMAP_PATH!!, ".")
    }

    private fun showDialog(file: File) {
        val filesButton = FilesButton()
        filesButton.setButtonVisibility(true, true, true, true, true, true)
        filesButton.setMessageText(getString(R.string.zh_file_message))
        filesButton.setMoreButtonText(getString(R.string.global_load))

        val filesDialog = FilesDialog(requireContext(), filesButton,
            { Tools.runOnUiThread { controlsListViewCreator?.refresh() } },
            controlsListViewCreator!!.fullPath, file
        )

        filesDialog.setCopyButtonClick { mPasteButton?.visibility = View.VISIBLE }

        filesDialog.setMoreButtonClick {
            val intent = Intent(requireContext(), CustomControlsActivity::class.java)
            val bundle = Bundle()
            bundle.putString(CustomControlsActivity.BUNDLE_CONTROL_PATH, file.absolutePath)
            intent.putExtras(bundle)

            startActivity(intent)
            filesDialog.dismiss()
        } //加载
        filesDialog.show()
    }

    private fun parseBundle() {
        val bundle = arguments ?: return
        mSelectControl = bundle.getBoolean(BUNDLE_SELECT_CONTROL, mSelectControl)
    }

    private fun bindViews(view: View) {
        mControlLayout = view.findViewById(R.id.control_layout)
        mOperateLayout = view.findViewById(R.id.operate_layout)

        mOperateView = view.findViewById(R.id.operate_view)

        mReturnButton = view.findViewById(R.id.zh_return_button)
        mImportControlButton = view.findViewById(R.id.zh_add_file_button)
        mAddControlButton = view.findViewById(R.id.zh_create_folder_button)
        mPasteButton = view.findViewById(R.id.zh_paste_button)
        mRefreshButton = view.findViewById(R.id.zh_refresh_button)
        mSearchSummonButton = view.findViewById(R.id.zh_search_button)
        mNothingTip = view.findViewById(R.id.zh_controls_nothing)

        mImportControlButton?.setContentDescription(getString(R.string.zh_controls_import_control))
        mAddControlButton?.setContentDescription(getString(R.string.zh_controls_create_new))

        controlsListViewCreator =
            ControlsListViewCreator(requireContext(), view.findViewById(R.id.zh_controls_list))

        mSearchViewWrapper = SearchViewWrapper(view, view.findViewById(R.id.zh_search_view))
        mSearchViewWrapper?.setAsynchronousUpdatesListener(object : SearchViewWrapper.SearchAsynchronousUpdatesListener {
            override fun onSearch(searchCount: TextView?, string: String?, caseSensitive: Boolean) {
                controlsListViewCreator?.searchControls(searchCount, string, caseSensitive)
            }
        })
        mSearchViewWrapper?.setShowSearchResultsListener(object : SearchViewWrapper.ShowSearchResultsListener {
            override fun onSearch(show: Boolean) {
                controlsListViewCreator?.setShowSearchResultsOnly(show)
            }
        })

        mPasteButton?.setVisibility(if (PasteFile.getInstance().pasteType != null) View.VISIBLE else View.GONE)

        ZHTools.setTooltipText(mReturnButton, mReturnButton?.contentDescription)
        ZHTools.setTooltipText(mImportControlButton, mImportControlButton?.contentDescription)
        ZHTools.setTooltipText(mAddControlButton, mAddControlButton?.contentDescription)
        ZHTools.setTooltipText(mPasteButton, mPasteButton?.contentDescription)
        ZHTools.setTooltipText(mSearchSummonButton, mSearchSummonButton?.contentDescription)
        ZHTools.setTooltipText(mRefreshButton, mRefreshButton?.contentDescription)
    }

    private fun startNewbieGuide() {
        if (NewbieGuideUtils.showOnlyOne(TAG)) return
        val fragmentActivity = requireActivity()
        TapTargetSequence(fragmentActivity)
            .targets(
                NewbieGuideUtils.getSimpleTarget(fragmentActivity, mRefreshButton, getString(R.string.zh_refresh), getString(R.string.zh_newbie_guide_general_refresh)),
                NewbieGuideUtils.getSimpleTarget(fragmentActivity, mSearchSummonButton, getString(R.string.zh_search), getString(R.string.zh_newbie_guide_control_search)),
                NewbieGuideUtils.getSimpleTarget(fragmentActivity, mImportControlButton, getString(R.string.zh_controls_import_control), getString(R.string.zh_newbie_guide_control_import)),
                NewbieGuideUtils.getSimpleTarget(fragmentActivity, mAddControlButton, getString(R.string.zh_controls_create_new), getString(R.string.zh_newbie_guide_control_create)),
                NewbieGuideUtils.getSimpleTarget(fragmentActivity, mReturnButton, getString(R.string.zh_return), getString(R.string.zh_newbie_guide_general_close)))
            .start()
    }

    override fun slideIn(animPlayer: AnimPlayer) {
        animPlayer.apply(AnimPlayer.Entry(mControlLayout!!, Animations.BounceInDown))
            .apply(AnimPlayer.Entry(mOperateLayout!!, Animations.BounceInLeft))
            .apply(AnimPlayer.Entry(mOperateView!!, Animations.FadeInLeft))
    }

    override fun slideOut(animPlayer: AnimPlayer) {
        animPlayer.apply(AnimPlayer.Entry(mControlLayout!!, Animations.FadeOutUp))
            .apply(AnimPlayer.Entry(mOperateLayout!!, Animations.FadeOutRight))
    }
}

